### ISSUE TITLE
[BUG]: buffer tank & soft start temperature correction

### ISSUE THREAD

**Issue by sailor25462** [edterbak/NodeRed_Heishamon_control#177]
**Describe the bug**
The soft start function shows unexpected behavior when a buffer tank is installed and activated.
Instead of (temporarily) reducing the target water temperature, the set target is increased.

**To Reproduce**
To reproduce the behavior, a buffer tank must be activated in the Panasonic control panel and heishamon board (Top 99). 
Top 113 "Buffer_Tank_Delta" must be set to some value >0K, as the Panasonic system will increase the output temp by the set delta. Required sensors (buffer tank temp, heating circuit temp, outside temp) have to be set properly to make the Panasonic system accept the buffer tank and keep it enabled.
Check in the node red dashboard, that buffer tank shows "enabled".
<img width="722" alt="Bildschirmfoto 2024-12-01 um 11 29 41" src="https://github.com/user-attachments/assets/5963ad7d-2ccd-4f48-89bc-e2197d3711bd">

Enable soft start and see the correction value turning "+$k" instead of "-$k". 

**Expected behavior**
Without  buffer tank being enable, soft start will reduce (temporarily) the target temperature. With the buffer tank being enabled, target temperature should still be reduced for startup.

**Screenshots**
<img width="342" alt="Bildschirmfoto 2024-12-01 um 11 25 01" src="https://github.com/user-attachments/assets/c2aa452d-587f-4d28-ad65-c80390be6ec3">

**Comment by geduxas** [comment]
I am trying to think, why it's happening, now can't test it by myself, but as i understand for softstart @edterbak you're reducing compensation curve setpoint? Right? So referring to screenshot, softstart is set to +1? If in start it was at 0, so it will naturaly get up.. 

Also @edterbak by what factor do you decide to increase or decrease set pont? Do you calculate it somehow from compensation curve? And comparing it with main target temperature? I'm truly not familiar with this logic, and didn't use soft start :)

So main target temperature will be compensation curve setpoint+shift value + buffer tank delta+2° without buffer tank target temperature is only curve setpoint+shift value.
 Maybe you will find something in softstart logic from this.

**Comment by edterbak** [comment]
Thank you geduxas !!!

I really need guidance here. So thank you.

The softstart function tries to catch a start of a heat cycle. When the compressor starts, it reduces the target as much as possible. 
This is reasonable straight forward when Heating mode = DIRECT. Just manipulate the current setpoint.

In case of the Panasonic Compensation curve is used, the SHIFT function is here, but only when NOT using a thermostat. 
The internal shift function allows us to push the curve up or down by x (between -5/+5)

When a thermostat is used, it is a bit more tricky be cause the SHIFT function does not work.  BUT>. of course this does not stop me. 

I actually grab the outside temperature, grab the temperatures from the compensation curves. the waterhigh/low and outsidehigh/low temperature when the shift should be 0. 

The RTC function calculates the correction required and add that to the curve. 
With the SET16 command, I can send a new curve to the pump. With this, I shift the curve manually.


But I think THIS is key!!!
>target temperature will be compensation curve setpoint+shift value + buffer tank delta+2°

**Would the situation be solved if I reduce the setpoint by (2+tank_delta) in case a buffer tank is used?**

**Comment by sailor25462** [comment]
Hey guys, sorry for not having answered to this one here, as I have apparently missed the notification. sorry.

I'm running 25.02 stable now and recently stumbled also over some weird temperature result in the correction, for which I will issue a separate report.

Now, regarding the buffer tank issue. Naturally, I haven't been using the heat pump over summer time, but have now re-started to heat and will have a look into it again.
In the meantime I had helped myself by setting the delta to 0°.

I think it is possible to simulate the buffer tank, as you just need to make sure you send frequently temperatures for the buffer tank and the temp in the heating circuit. without those data, the heat pump will throw an error internally, reset and deactivate the buffer tank. So, just in case you want to enable it by yourself, I am just pushing my sensor readings (which your can fake) via node red to the Heishamon every 30 seconds.

With the buffer tank being enable, the heat pump internally add the delta (normally 5°C) to the set temperature. But this is not shown at any register, but only done "silently". 
My guess is that in order to get the soft start right, you need to read the delta from the settings and consider it in your correction.

**Comment by geduxas** [comment]
Yes buffer tank delta adds to bigest zone setpoint as target final temperature, it should not be set from node-red or any other sources, pump calculates and adds extra to final desired temperature, and for rest operations it should be treated as additional temperature to zone setpoints.. 

So if you're zone set point is 35°, and delta is 5K so pump will show final setpoint temperature as 40° (35+5)
So 40, should not be treated as zone set point, because it's still is 35°. Only difference is that, buffer tank temperature will be 40°, and naturally from buffer tank you should have mixing valve which will drive it's desired temperature to 35°.

**Comment by sailor25462** [comment]
ok, maybe there's a misunderstanding on my side then. Let's recap the installation manual and the requirements for using a buffer tank.

<img width="883" height="473" alt="Image" src="https://github.com/user-attachments/assets/9cf80557-f8ef-4de2-a4dc-171f3d54ad35" />

<img width="888" height="405" alt="Image" src="https://github.com/user-attachments/assets/fce3affb-d6b5-4ec1-a18a-f6fa48bfd8eb" />

<img width="897" height="527" alt="Image" src="https://github.com/user-attachments/assets/29b1f6b0-5df5-4183-adee-32fee36624c6" />


Looking at the error scheme, with the tank sensor (buffer temp) installed, also a zone water temp sensor is mandatory as well.
Without updating this data every 30 seconds, the pump will actually run into error and reset.
What confused me was, that the zone water temp sensor cannot be found in the schematic drawing for a single zone installation at all. However, if you look into 2 zone installations, the mandatory sensor is indeed shown (as #10).

<img width="926" height="1134" alt="Image" src="https://github.com/user-attachments/assets/17629841-549e-406a-8375-b4500015b278" />

So, what I actually did was to install an ESP32 with a temperature sensor, attached to the heating circuit at the pipe right after the buffer tank and secondary circulation pump, so it will show the true temperature in the heating circuit. The measured value is injected to SetZ1WaterTemp.
As said, without that, the pump runs into error and deactivates the optional circuit board (as well as the buffer).

**Comment by sailor25462** [comment]
In this regard, I think all is working as it should. 
All running pretty smoothly, except the softstart. 
When I maintain a buffer delta, the pump runs just not as „low“ as without, but using +5k or more in the softstart settings instead. 
Do you think there‘s something you could improve? 
Or is it not a bug but a „feature“, then I‘ll probably just leave the buffer delta at 0k

**Comment by edterbak** [comment]
is there still an mqtt command spam?

If so, I need to understand exactly the issue and fix it
